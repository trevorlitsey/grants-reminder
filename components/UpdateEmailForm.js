import { string, func } from 'prop-types';
import validator from 'email-validator';

import FormAlert from './FormAlert';

class UpdateEmailForm extends React.PureComponent {

	constructor(props) {
		super(props);

		this.setError = this.setAlert.bind(this);
		this.handleChange = this.handleChange.bind(this);
		this.handleSubmit = this.handleSubmit.bind(this);

		this.state = {
			newEmail: '',
			alertMsg: '',
		}
	}

	setAlert(alertMsg) {
		this.setState({ alertMsg });
		setTimeout(() => {
			this.setState({ alertMsg: '' });
		}, 3000);
	}

	handleChange(e) {
		this.setState({
			newEmail: e.target.value,
		})
	}

	handleSubmit(e) {
		e.preventDefault();
		const { newEmail } = this.state;
		const emailIsValid = validator.validate(newEmail);
		if (!emailIsValid) {
			return this.setAlert(`hmm that doesn't look like a valid email`);
		} else {
			this.setAlert(`success!`);
			// update email
			return this.props.updateEmail(this.state.newEmail)
		}
	}

	render() {
		return (
			<div>
				<form onSubmit={this.handleSubmit}>
					<input
						type="text"
						name="newEmail"
						placeholder={`${this.props.email} ...`}
						value={this.state.newEmail}
						onChange={this.handleChange}
					/>
					<input className="submit" type="submit" value="Update Email" />
				</form>
				<FormAlert message={this.state.alertMsg} />

				<style jsx>{`

					form {
						margin-top: 20px;
						margin-bottom: 0;
						display: flex;
					}
					
					input {
						margin: 0 2px;
					}
					
					.submit {
						padding: 0 10px;
					}
					
				`}</style>
			</div>
		)
	}
}

UpdateEmailForm.propTypes = {
	email: string.isRequired,
	updateEmail: func.isRequired,
}


export default UpdateEmailForm;