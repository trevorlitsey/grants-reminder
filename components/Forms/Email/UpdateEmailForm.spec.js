import React from 'react';
import { shallow } from 'enzyme';

import UpdateEmailForm from './UpdateEmailForm';

// setError(alertMsg) {
// 	this.setState({ alertMsg });
// 	setTimeout(() => {
// 		this.setState({ alertMsg: '' });
// 	}, 3000);
// }

// handleSubmit(e) {


describe('UpdateEmailForm', () => {

	it('should render without crashing', () => {
		const wrapper = shallowRenderUpdateEmailForm();
		expect(wrapper).toMatchSnapshot();
	})

	it('should handle input change and update state', () => {
		const instance = shallowRenderUpdateEmailForm().instance();
		const newEmail = 'new@verynew.com'
		const event = makeOnChangeEvent('email', newEmail)
		instance.handleChange(event);
		expect(instance.state.newEmail).toEqual(newEmail);
	})

	it('should throw an error if email is invalid', () => {
		const event = {
			preventDefault: () => { },
		}
		const instance = shallowRenderUpdateEmailForm().instance();
		const newEmail = 'soooonew.com'
		instance.setState({ newEmail })
		instance.handleSubmit(event);
		expect(instance.state.alertMsg).toEqual(`hmm that doesn't look like a valid email`);
	})

	it('display SUCCESS if email is valid', () => {
		const event = {
			preventDefault: () => { },
		}
		const instance = shallowRenderUpdateEmailForm().instance();
		const newEmail = 'soooonew@gmail.com'
		instance.setState({ newEmail })
		instance.handleSubmit(event);
		expect(instance.state.alertMsg).toEqual(`success!`);
	})

})

function shallowRenderUpdateEmailForm(props = {}) {
	const propsToUser = {
		email: 'test@test.com',
		updateEmail: () => { },
		...props,
	}
	return shallow(<UpdateEmailForm {...propsToUser} />)
}

function makeOnChangeEvent(name, value) {
	const event = {
		preventDefault: () => { },
		target: {
			name,
			value,
		}
	}
	return event;
}