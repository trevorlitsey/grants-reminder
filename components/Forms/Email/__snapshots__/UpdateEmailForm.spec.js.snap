// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateEmailForm should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UpdateEmailForm
    email="test@test.com"
    updateEmail={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          className="jsx-2624128934"
          onSubmit={[Function]}
        >
          <input
            className="jsx-2624128934"
            name="newEmail"
            onChange={[Function]}
            placeholder="test@test.com ..."
            type="text"
            value=""
          />
          <input
            className="jsx-2624128934 submit"
            type="submit"
            value="Update Email"
          />
        </form>,
        <FormAlert
          message=""
        />,
        <JSXStyle
          css="form.jsx-2624128934{margin-top:20px;margin-bottom:0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}input.jsx-2624128934{margin:0 2px;}.submit.jsx-2624128934{padding:0 10px;}"
          styleId="2624128934"
        />,
      ],
      "className": "jsx-2624128934",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="jsx-2624128934"
              name="newEmail"
              onChange={[Function]}
              placeholder="test@test.com ..."
              type="text"
              value=""
            />,
            <input
              className="jsx-2624128934 submit"
              type="submit"
              value="Update Email"
            />,
          ],
          "className": "jsx-2624128934",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "jsx-2624128934",
              "name": "newEmail",
              "onChange": [Function],
              "placeholder": "test@test.com ...",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "jsx-2624128934 submit",
              "type": "submit",
              "value": "Update Email",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "css": "form.jsx-2624128934{margin-top:20px;margin-bottom:0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}input.jsx-2624128934{margin:0 2px;}.submit.jsx-2624128934{padding:0 10px;}",
          "styleId": "2624128934",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            className="jsx-2624128934"
            onSubmit={[Function]}
          >
            <input
              className="jsx-2624128934"
              name="newEmail"
              onChange={[Function]}
              placeholder="test@test.com ..."
              type="text"
              value=""
            />
            <input
              className="jsx-2624128934 submit"
              type="submit"
              value="Update Email"
            />
          </form>,
          <FormAlert
            message=""
          />,
          <JSXStyle
            css="form.jsx-2624128934{margin-top:20px;margin-bottom:0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}input.jsx-2624128934{margin:0 2px;}.submit.jsx-2624128934{padding:0 10px;}"
            styleId="2624128934"
          />,
        ],
        "className": "jsx-2624128934",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="jsx-2624128934"
                name="newEmail"
                onChange={[Function]}
                placeholder="test@test.com ..."
                type="text"
                value=""
              />,
              <input
                className="jsx-2624128934 submit"
                type="submit"
                value="Update Email"
              />,
            ],
            "className": "jsx-2624128934",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "jsx-2624128934",
                "name": "newEmail",
                "onChange": [Function],
                "placeholder": "test@test.com ...",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "jsx-2624128934 submit",
                "type": "submit",
                "value": "Update Email",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "css": "form.jsx-2624128934{margin-top:20px;margin-bottom:0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}input.jsx-2624128934{margin:0 2px;}.submit.jsx-2624128934{padding:0 10px;}",
            "styleId": "2624128934",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
